ocaml/testsuite/tests/typing-modes/portable-contend.ml

```diff
@@ -0,0 +1,249 @@
+(* TEST
+   * expect *)
+
+type r = {mutable a : string; b : string}
+
+(* TESTING records *)
+
+(* Reading/writing mutable field from contended record is rejected. Also note
+    that the mutation error precedes type error. *)
+let foo (r @ contended) = r.a <- 42
+[%%expect{|
+type r = { mutable a : string; b : string; }
+Line 7, characters 26-27:
+7 | let foo (r @ contended) = r.a <- 42
+                              ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+let foo (r @ contended) = r.a
+[%%expect{|
+Line 1, characters 26-27:
+1 | let foo (r @ contended) = r.a
+                              ^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+(* reading immutable field from contended record is fine *)
+let foo (r @ contended) = r.b
+[%%expect{|
+val foo : r @ contended -> string @ contended = <fun>
+|}]
+
+(* Force top level to be uncontended and nonportable *)
+let r @ contended = "hello"
+[%%expect{|
+Line 1, characters 4-27:
+1 | let r @ contended = "hello"
+        ^^^^^^^^^^^^^^^^^^^^^^^
+Error: Found a contended value where a uncontended value was expected
+|}]
+
+let x @ portable = "world"
+
+let y @ portable = x
```

Comment:
Good point. What's happening here is that, we are at top-level, which enables a special behaviour:
- When type checking a module NOT top-level, we type check all the values without mode restriction, and in the end requiring them to be submode of `legacy`. That is, if we put this example inside a `module`, there won't be any error at all, because both `x` and `y` will be `portable` which is submode of `nonportable` as required.
- When type checking at top-level, we type check all the values at `legacy`. That means the `x` in the example will be required to be `nonportable`, and made available in the environment at `nonportable`, which causes the error.

I'm not 100% sure about why this top-level behaviour, some comment about that in `typemod.ml` (search for `force_toplevel`).

Reply:
That code seems like it has more to do with locality than modes in general. It looks like this can be fixed.

Reply:
I think the reason for the special treatment is simply that top-level phrases are type-checked loosely and differently, and it's  safe and simple to assume everything is legacy.

Reply:
But maybe we shouldn't test top-level at all - it's rarely used. I will put both `x` and `y` into a module, which is worth testing.

Reply:
Is this testing the top-level of the REPL? Or is this treatment true of type-level definitions in any file? I think it's probably the former (and then I agree -- though we should document), but I wanted to double-check.

Reply:
Yes I believe those `expect` tests are backed by top-level of REPL, not file-level modules.

I suggest that we test both, because some people do use REPL sometimes.

Reply:
In the end, I keep those tests unchanged. Because the fact that "values in a module is always legacy" is tested separately by `module.ml` already.